Redefine methods with correct visibility upfront

paracycle opened the pull request:
### Motivation
<!-- Why make this change? Describe the problem, not the solution. This can also be a link to an issue. -->
Fixes #5786

Sorbet runtime updates the visibility of the methods it is replacing but sometimes that happens too late. For example, if user code has a `method_added` hook and checks the visibility of the method that's added in the hook, it would always see `public` when Sorbet replaces a `private` or `protected` method. This was happening because the setting of the visibility was happening after the call to `define_method` and by then it was too late.

Instead of setting the visibility after defining the method, this PR encapsulates a method definition operation inside a method named `def_with_visibility` on the `T::Private::ClassUtils` module, which does a `module_exec` on the passed in module so that it can create a `public`/`protected`/`private` section and only then define the method. This ensures that the method is defined with the expected visibility from the beginning.

### Test plan
<!-- If you did not write tests for this change, replace the message below explaining why not. Why we should be confident this change is correct? If you changed the website, please include a screenshot of the proposed changes. -->

See included automated tests. Added a test to verify that `method_added` hooks see methods that are replaced by Sorbet runtime with the expected visibility.

jez commented:
Thanks, I've kicked off a build on Stripe's codebase.

http://go/builds/bui_MDp1brTspM2YrB

jez reviewed (changes requested):
It looks like you have some test failures? Feel free to re-request review when they pass, I expect the change itself is fine but I'll look more closely when the tests pass.

paracycle commented:
> It looks like you have some test failures? Feel free to re-request review when they pass, I expect the change itself is fine but I'll look more closely when the tests pass.

It is mostly that I had missed the comment on top of the `call_validation` files and updated them manually. Now I fixed the generator, so it should be fine.

paracycle commented on gems/sorbet-runtime/lib/types/private/class_utils.rb:
```diff
@@ -69,6 +69,27 @@ def to_s
+        define_method(name, &block)
+      end
+
+      if block && block.arity < 0 && respond_to?(:ruby2_keywords, true)
+        ruby2_keywords(name)
+      end
```
I encapsulated this `ruby2_keywords` call in this method as well, but I am not sure if that's the best thing to do. The `define_method` calls originating from `call_validation` were never changing the `ruby2_keywords` status of methods, but I am not sure if that was because it was forgotten or if it was intentional.

I'd be ok with moving this part out back to the original call sites if needed.

jez replied:
Honestly I think you'd be better at telling me where this `ruby2_keywords` call has to go, as Stripe's codebase is still using Ruby 2.7, and so is unlikely to turn up any problems from incorrectly placement of this.

jez commented:
> Thanks, I've kicked off a build on Stripe's codebase.
> 
> http://go/builds/bui_MDp1brTspM2YrB

There was one problem, which was actually from a previous change to sorbet-runtime that we simply forgot to update to.

Fixing that one problem led to this build http://go/builds/bui_MDpByg3Qlsga8x which passed.

paracycle commented:
Ok, the build is green now!
